packages/@okta/vuepress-site/code/java/spring_security_saml/index.md

```diff
@@ -7,217 +7,372 @@ excerpt: >-
 
-Here's what it should look like: ![App Running](/img/spring-security-saml-intro.png "App Running")
+   ```shell
+   ./gradlew bootRun
+   ```
 
-## Configure Okta to work with Spring Security SAML
+1. Open `http://localhost:8080` in your favorite browser and sign in with the credentials you used to create your account. You should see a successful result in your browser.
+
+   <div class="three-quarter border">
+
+   ![Login success](/img/saml/spring-security-saml-login.png)
+
+   </div>
+
+If you try to sign out, it won't work. You'll fix that in the next section.
+
+### Add a sign out feature
+
+Spring Security's SAML support has a [sign out feature](https://docs.spring.io/spring-security/reference/servlet/saml2/logout.html) that requires configuration.
+
+1. Edit your application on Okta and navigate to **General** > **SAML Settings** > **Edit**.
+
+1. Continue to the **Configure SAML** step and **Show Advanced Settings**. Before you can enable single logout, you'll have to create and upload a certificate to sign the outgoing logout request.
+
+1. You can create a private key and certificate using OpenSSL. Answer at least one of the questions with a value, and it should work.
+
+   ```shell
+   openssl req -newkey rsa:2048 -nodes -keyout local.key -x509 -days 365 -out local.crt
+   ```
+
+1. Copy the generated files to your app's `src/main/resources` directory. Configure `signing` and `singlelogout` in `application.yml`:
+
+   ```yaml
+   spring:
+     security:
+       saml2:
+         relyingparty:
+           registration:
+             okta:
+               signing:
+                 credentials:
+                   - private-key-location: classpath:local.key
+                     certificate-location: classpath:local.crt
+               singlelogout:
+                 binding: POST
+                 response-url: "{baseUrl}/logout/saml2/slo"
+               assertingparty:
+                 metadata-uri: <your-metadata-uri>
+   ```
+
+1. Upload the `local.crt` to your Okta app. Select *Enable Single Logout* and use the following values:
+
+   * Single Logout URL: `http://localhost:8080/logout/saml2/slo`
+   * SP Issuer: `http://localhost:8080/saml2/service-provider-metadata/okta`
+
+1. Finish configuring your Okta app, restart your Spring Boot app, and the logout button should work.
+
+   <div class="three-quarter border">
+
+   ![Sign out success](/img/saml/spring-security-saml-logout-success.png)
+
+   </div>
+
+### Customize authorities with Spring Security SAML
 
-Before we can configure Spring Security SAML, we need to set up an application in Okta that connects to Spring Security SAML.
+When you sign in, the resulting page shows that you have a `ROLE_USER` authority. However, when you assigned users to the app, you gave access to `Everyone`. You can configure your SAML app on Okta to send a user's groups as an attribute, and add other attributes like name and email.
```

Comment:
This paragraph doesn't make sense to me. The heading says "Customize authorities" but most of this paragraph is about adding attributes to something that isn't specified, so where is the connection between Okta attributes and whatever authorities are? I think you need to explain that.

Reply:
SAML calls everything an attribute. In this example, I'm converting Okta's groups to Spring Security authorities. Java developers that are familiar with Spring Security should understand this, IMO.

Reply:
OK - I'll leave that to your judgment